### DISCUSSION TITLE
Generalizing Canonical allele and Categorical CNV to handle Function/Expression variants

### DISCUSSION THREAD

**Discussion by DanielPuthawala** [ga4gh/cat-vrs#16] [General]
Currently there are two issues with the current structure of the schema that I would like to address.

First, the CanonicalAllele and Protein SequenceConsequence both exist and are more or less identical, but are simply meant to describe sequence variants at two different levels, genomic and protein, respectively. While classes are cheap, DRY. I want to explore the option to subsume these two classes under a more general SequenceVariant class (name can be contested) that keeps the canonicalized properties of both, while differentiating genomic from protien variants by some sort of required level: one of {"genomic","protein"} property.

Second, while sequence variants are handled by CanonicalAllele or the new SequenceVariant class , and CNVs are handled by the CategoricalCNV class (subject to ay changes as a result of issue-11), there is a coherent class of variants not covered so far: Function Variants. Here's some suCatVars in this category:

Overexpression Variants
Unerexpression Variants
Gain-of-Function (GoF) Variants
Loss-of-Function (LoF) Variants
Loss Variants
Amplification Variants
Nonsense Variants (?)

These variants all entail some sort of assayed quantitative change, often but not always associated with some sort of sequence change. That is, capturing this class of CatVars requires properties from both the CategoricalCNV class and the SequenceVariant class. We could either create a third class that just happens to contain the same classes as both of those, or I propose we make a third class that is a generalization of the notion of quantity change and sequence change variants, subsuming both CategoricalCNV and SequenceVariant into a larger class, which, by way of optional properties, can handle CNVs, sequence variants, AND intermediate cases like the function variants described above.

**Comment by DanielPuthawala** [comment]
By the way, this will correspond to what I'm doing on the Issue-14 branch.

**Comment by DanielPuthawala** [comment]
Maybe we should iron out the details on CNVs before we expand on that.

**Reply by larrybabb** [reply]
I'm not sure I agree with the statement that canonical allele and protein expressions are almost identical. Canonical allele from my experience so far has always been about defining a set of variants that derive from one genomic dna location and change. Adding other functional or expression attributes makes no sense to canonical allele afaic. But I'm willing to hear the counter argument of how these two types can be defined and compared in an apples-to-apples manner.

**Comment by DanielPuthawala** [comment]
I thikn we've moved past the point in development where this discussion is outdated, so I'm closing this discussion.

**Comment by melissacline** [comment]
@DanielPuthawala, at the risk of extending a closed discussion, I'm interested in the plans for overexpression / underexpression variants.  Are these outside of the scope of Cat-VRS for now?
